src/classes/PMT_Payment_TDTM.cls

```diff
@@ -44,18 +103,51 @@ public class PMT_Payment_TDTM extends TDTM_Runnable {
-        if (oldlist != null) 
-           mapIdOppOld = new map<id, Opportunity>((list<Opportunity>) oldlist);
-                       
-        DmlWrapper dmlWrapper = new dmlWrapper();
-        PMT_PaymentCreator pc = new PMT_PaymentCreator(mapIdOppNew, mapIdOppOld, triggerAction, dmlWrapper); 
+    public static DmlWrapper runForOpportunities(List<Opportunity> newlist, List<Opportunity> oldlist,
+            TDTM_Runnable.Action triggerAction) {
+
+        Map<Id, Opportunity> mapIdOppNew = (newlist != null ? new Map<Id, Opportunity>(newlist) : new Map<Id, Opportunity>());
+        Map<Id, Opportunity> mapIdOppOld = (oldlist != null ? new Map<Id, Opportunity>(oldlist) : new Map<Id, Opportunity>());
+
+        DmlWrapper dmlWrapper = new DmlWrapper();
+
+        // Create new Payments as necessary
+        PMT_PaymentCreator pc = new PMT_PaymentCreator(mapIdOppNew, mapIdOppOld, triggerAction, dmlWrapper);
+
+        //if currency has changed, reset payment currencies
+        if (triggerAction == TDTM_Runnable.Action.AfterUpdate && UTIL_Currency.getInstance().isMultiCurrencyOrganization()) {
+            List<Opportunity> listOppsForProcessing = new List<Opportunity>();
+            for (integer i = 0; i < newlist.size(); i++) {
+                Opportunity opp = newlist[i];
+                if (opp.get('CurrencyIsoCode') != oldlist[i].get('CurrencyIsoCode')) {
+                    listOppsForProcessing.add(opp);
+                }
+            }
+            if (listOppsForProcessing.size() > 0) {
+                setPaymentCurrencyFromOpportunity(listOppsForProcessing, dmlWrapper);
+            }
+        }
         return dmlWrapper;
     }
+
+    /*******************************************************************************************************
+    * @description Updates an opportunity's payment record when currency on the Opportunity is changed.
+    * @param listOpps A list of opportunities whose currency has changed.
+    * @return void This method modifies the dmlWrapper var and does not return anything
+    ********************************************************************************************************/
+    private static void setPaymentCurrencyFromOpportunity(list<Opportunity> listOpps, DmlWrapper dmlWrapper){
+
+        // Get a list of the Opportunity Id's
+        Set<Id> oppIds = (new Map<Id, Opportunity>(listOpps)).keySet();
+
+        // Get a list of the Payment records by Opportunity and update the CurrencyIsoCode value
+        String soql = 'SELECT Id, CurrencyIsoCode, npe01__Opportunity__r.CurrencyIsoCode FROM npe01__OppPayment__c ' +
+            'WHERE npe01__Opportunity__c IN :oppIds ORDER BY npe01__Opportunity__c';
+        for (npe01__OppPayment__c pmt : database.query(soql)) {
+            Opportunity opp = (Opportunity)pmt.getSObject('npe01__Opportunity__r');
+            pmt.put('CurrencyIsoCode', opp.get('CurrencyIsoCode'));
+            dmlWrapper.objectsToUpdate.add(pmt);
```

Comment:
are there scenarios where the payments will already have the correct isocode, and we should only update them if they need it?  just trying to keep down unnecessary dml if possible.

Reply:
That probably shouldn't happen. This method would only be called if the Opportunity CurrencyIsoCode gets changed, so if the values are kept in sync, and the Opportunity value is changed, then the Payment would not have the right currency until this runs. Basically every record _should_ be updated.